rust_src/src/remacs_sys.rs

```diff
@@ -1597,18 +1597,44 @@ extern "C" {
     pub fn fset_selected_window(frame: *mut Lisp_Frame, window: LispObject);
 }
 
+#[repr(C)]
+pub union display_info {
+    pub tty: *const tty_display_info,
+    pub x: *const x_display_info,
+    pub w32: *const w32_display_info,
+    pub ns: *const ns_display_info,
+}
+
 #[repr(C)]
 pub struct terminal {
     pub header: Lisp_Vectorlike_Header,
+    pub kboard: kboard,
+    pub display_info: display_info,
 }
 
+#[repr(C)]
+#[derive(PartialEq)]
+pub struct kboard;
+
+#[repr(C)]
+pub struct tty_display_info;
+
+#[repr(C)]
+pub struct x_display_info;
+
+#[repr(C)]
+pub struct w32_display_info;
+
+#[repr(C)]
+pub struct ns_display_info;
```

Comment:
Can these safely be zsts? I suppose if we never have a vector of them, then it will be ok. You'll either need to fill them out (hard), or tag them with `#![allow(improper_ctypes)]` (easy).

Reply:
Sounds sensible :)

Could you add a comment explaining that we're only using improper_ctypes whilst we don't have full struct declarations for frames?